src/NServiceBus.RavenDB/SessionManagement/RavenDbStorageSession.cs

```diff
@@ -16,24 +18,39 @@ public RavenDbStorageSession()
 
         protected override void Setup(FeatureConfigurationContext context)
         {
+            var store = DocumentStoreManager.GetDocumentStore<StorageType.Sagas>(context.Settings);
+
+            IOpenRavenSessionsInPipeline sessionCreator;
+
             // Check to see if the user provided us with a shared session to work with before we go and create our own to inject into the pipeline
-            var getSessionFunc = context.Settings.GetOrDefault<Func<IDocumentSession>>(RavenDbSettingsExtensions.SharedSessionSettingsKey);
+            var getSessionFunc = context.Settings.GetOrDefault<Func<IDictionary<string, string>, IDocumentSession>>(RavenDbSettingsExtensions.SharedSessionSettingsKey);
+            var getSessionFuncObsolete = context.Settings.GetOrDefault<Func<IDocumentSession>>(RavenDbSettingsExtensions.SharedSessionSettingsKey + ".Obsolete");
+
             if (getSessionFunc != null)
             {
-                context.Container.ConfigureComponent<ProvidedSessionBehavior>(DependencyLifecycle.InstancePerCall)
-                    .ConfigureProperty(x => x.GetSession, getSessionFunc);
-                context.Pipeline.Register<ProvidedSessionBehavior.Registration>();
-                return;
+                sessionCreator = new OpenRavenSessionByCustomDelegate(getSessionFunc);
+            }
+            else if (getSessionFuncObsolete != null)
+            {
+                sessionCreator = new OpenRavenSessionByCustomDelegate(getSessionFuncObsolete);
             }
+            else
+            {
+               
+                var storeWrapper = new DocumentStoreWrapper(store);
 
-            var store = DocumentStoreManager.GetDocumentStore<StorageType.Sagas>(context.Settings);
+                var dbNameConvention = context.Settings.GetOrDefault<Func<IMessageContext, string>>("RavenDB.SetMessageToDatabaseMappingConvention");
+                sessionCreator = new OpenRavenSessionByDatabaseName(storeWrapper, dbNameConvention);
+            }
 
             context.Container.ConfigureComponent<Installer>(DependencyLifecycle.InstancePerCall)
                 .ConfigureProperty(c => c.StoreToInstall, store);
 
             context.Container.ConfigureComponent<RavenSessionProvider>(DependencyLifecycle.InstancePerCall);
             context.Container.RegisterSingleton<IDocumentStoreWrapper>(new DocumentStoreWrapper(store));
             context.Pipeline.Register<OpenSessionBehavior.Registration>();
+
+            context.Container.RegisterSingleton(sessionCreator);
```

Comment:
do I get it right that `sessionCreator` cannot be `null` in any way. If user provides a customization delegate we use that, otherwise we fall back to default.

Reply:
@mauroservienti You're correct. The `OpenRavenSessionByDatabaseName` implementation will just use the default documentStore if the `dbNameConvention` func retrieved on Line 42 is null.